unison-cli/src/Unison/Share/Sync.hs

```diff
@@ -0,0 +1,764 @@
+import qualified Servant.Client as Servant (ClientEnv, ClientM, client, hoistClient, mkClientEnv, runClientM)
+import U.Codebase.HashTags (CausalHash (..))
+import qualified U.Codebase.Sqlite.Branch.Format as NamespaceFormat
+import qualified U.Codebase.Sqlite.Causal as Causal
+import qualified U.Codebase.Sqlite.Decl.Format as DeclFormat
+import qualified U.Codebase.Sqlite.Entity as Entity
+import U.Codebase.Sqlite.LocalIds (LocalIds' (..))
+import qualified U.Codebase.Sqlite.Patch.Format as PatchFormat
+import qualified U.Codebase.Sqlite.Queries as Q
+import U.Codebase.Sqlite.TempEntity (TempEntity)
+import qualified U.Codebase.Sqlite.TempEntity as TempEntity
+import qualified U.Codebase.Sqlite.Term.Format as TermFormat
+import U.Util.Base32Hex (Base32Hex)
+import qualified U.Util.Hash as Hash
+import Unison.Auth.HTTPClient (AuthorizedHttpClient)
+import qualified Unison.Auth.HTTPClient as Auth
+import Unison.Prelude
+import qualified Unison.Sqlite as Sqlite
+import qualified Unison.Sync.API as Share (api)
+import qualified Unison.Sync.Types as Share
+import qualified Unison.Sync.Types as Share.RepoPath (RepoPath (..))
+import Unison.Util.Monoid (foldMapM)
+import qualified Unison.Util.Set as Set
+
+------------------------------------------------------------------------------------------------------------------------
+-- Push
+
+-- | An error occurred while pushing code to Unison Share.
+data CheckAndSetPushError
+  = CheckAndSetPushErrorHashMismatch Share.HashMismatch
+  | CheckAndSetPushErrorNoWritePermission Share.RepoPath
+  | CheckAndSetPushErrorServerMissingDependencies (NESet Share.Hash)
+
+-- | Push a causal to Unison Share.
+-- FIXME reword this
+checkAndSetPush ::
+  -- | The HTTP client to use for Unison Share requests.
+  AuthorizedHttpClient ->
+  -- | The Unison Share URL.
+  BaseUrl ->
+  -- | SQLite connection, for reading entities to push.
+  Sqlite.Connection ->
+  -- | The repo+path to push to.
+  Share.RepoPath ->
+  -- | The hash that we expect this repo+path to be at on Unison Share. If not, we'll get back a hash mismatch error.
+  -- This prevents accidentally pushing over data that we didn't know was there.
+  Maybe Share.Hash ->
+  -- | The hash of our local causal to push.
+  CausalHash ->
+  IO (Either CheckAndSetPushError ())
+checkAndSetPush httpClient unisonShareUrl conn repoPath expectedHash causalHash = do
+  -- Maybe the server already has this causal; try just setting its remote path. Commonly, it will respond that it needs
+  -- this causal (UpdatePathMissingDependencies).
+  updatePath >>= \case
+    Share.UpdatePathSuccess -> pure (Right ())
+    Share.UpdatePathHashMismatch mismatch -> pure (Left (CheckAndSetPushErrorHashMismatch mismatch))
+    Share.UpdatePathMissingDependencies (Share.NeedDependencies dependencies) -> do
+      -- Upload the causal and all of its dependencies.
+      uploadEntities httpClient unisonShareUrl conn (Share.RepoPath.repoName repoPath) dependencies >>= \case
```

Comment:
Can we just call `checkAndSetPush` again after uploading dependencies? Or are you worried about the 'how did this happen' case causing an infinite loop?

Reply:
What would be the purpose of that?